File: app/src/main/java/com/example/myapplication/di/modules/authentication/AuthenticationKey.java
package com.example.myapplication.di.modules.authentication;

import androidx.lifecycle.ViewModel;
import com.example.myapplication.domain.authentication.AuthenticationSource;
import dagger.MapKey;

import java.lang.annotation.*;

@MapKey
@Documented
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface AuthenticationKey {
    AuthenticationSources value();
}

File: app/src/main/java/com/example/myapplication/di/modules/authentication/AuthenticationKey.kt
package com.example.myapplication.di.modules.authentication

import com.example.myapplication.domain.authentication.AuthenticationProvider
import dagger.MapKey
import kotlin.reflect.KClass

@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
@MapKey
internal annotation class AuthenticationKey(val value: KClass<out AuthenticationProvider>)
